### PR TITLE
Optimize large memory usage of InsertRecord by using vector instead of unordered_map if InsertRecord used in sealed segment

### PR DESCRIPTION
**Pull request by aoiasd**
related:https://github.com/milvus-io/milvus/issues/18764 https://github.com/milvus-io/milvus/issues/18845
Signed-off-by: aoiasd <zhicheng.yue@zilliz.com>

### CODE REVIEW

**Comment by aoiasd** [comment]
rerun ut

File: internal/core/src/segcore/InsertRecord.h [hunk 1]
```diff
@@ -86,6 +95,70 @@ class OffsetHashMap : public OffsetMap {
     std::unordered_map<T, std::vector<int64_t>> map_;
 };
 
+template <typename T>
+class OffsetOrderedArray : public OffsetMap {
+ private:
+    int
+    find(const T pk) const {
```
Comment (yah01) [inline internal/core/src/segcore/InsertRecord.h:102]: use `std::binary_search()`

File: internal/core/src/segcore/InsertRecord.h [hunk 2]
```diff
@@ -86,6 +95,70 @@ class OffsetHashMap : public OffsetMap {
...
+            else
+                r = m;
+        }
+        return r;
+    }
+
+ public:
+    std::vector<SegOffset>
+    find_with_timestamp(const PkType pk, Timestamp timestamp, const ConcurrentVector<Timestamp>& timestamps) const {
+        std::vector<SegOffset> res_offsets;
```
Comment (yah01) [inline internal/core/src/segcore/InsertRecord.h:117]: construct this vector with iterator after replace `find` with `std::binary_search()`

File: internal/core/src/segcore/InsertRecord.h [hunk 3]
```diff
@@ -86,6 +95,70 @@ class OffsetHashMap : public OffsetMap {
...
+            }
+        }
+        return res_offsets;
+    }
+
+    std::vector<SegOffset>
+    find_with_barrier(const PkType pk, int64_t barrier) const {
+        std::vector<SegOffset> res_offsets;
+        T pk_ = std::get<T>(pk);
+        for (int offset_id = find(pk_); offset_id < array_.size(); offset_id++) {
```
Comment (yah01) [inline internal/core/src/segcore/InsertRecord.h:133]: same as line 117

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/core/unittest/test_segcore.cpp [hunk 4]
```diff
@@ -88,28 +90,92 @@ TEST(OffsetMap, int64_t){
...
     schema->AddDebugField("fakevec", DataType::VECTOR_FLOAT, 16, knowhere::metric::L2);
     auto i64_fid = schema->AddDebugField("age", DataType::INT64);
     schema->set_primary_field_id(i64_fid);
+    auto record = milvus::segcore::InsertRecord(*schema, int64_t(32));
+    int n=10000000;
+    Timer timer;
+
+    for (int i = 1; i <= n; i++)
+        record.insert_pk(PkType(int64_t(i)), int64_t(i));
+    std::cout << "insert pks: " << timer.get_step_seconds() << " seconds" << std::endl;
```
Comment (jiaoew1991) [inline internal/core/unittest/test_segcore.cpp:105]: remove it

Comment (jiaoew1991) [reply internal/core/unittest/test_segcore.cpp:105]: the same as L111 L126 etc

File: internal/core/src/segcore/InsertRecord.cpp [hunk 5]
```diff
@@ -14,7 +14,7 @@
 
 namespace milvus::segcore {
 
-InsertRecord::InsertRecord(const Schema& schema, int64_t size_per_chunk)
+InsertRecord::InsertRecord(const Schema& schema, int64_t size_per_chunk, bool is_sealed)
```
Comment (jiaoew1991) [inline internal/core/src/segcore/InsertRecord.cpp:17]: sealed is a caller concept，pass a creator or type  is better

**Pull request merged**